KAFKA-19441: encapsulate MetadataImage in GroupCoordinator/ShareCoordinator

zzbennett opened the pull request:
The MetadataImage has a lot of stuff in it and it gets passed around in many places in the new GroupCoordinator. This makes it difficult to understand what metadata the group coordinator actually relies on and makes it too easy to use metadata in ways it wasn't meant to be used. 

This change encapsulate the MetadataImage in an interface (`CoordinatorMetadataImage`) that indicates and controls what metadata the group coordinator actually uses. Now it is much easier at a glance to see what dependencies the GroupCoordinator has on the metadata. Also, now we have a level of indirection that allows more flexibility in how the GroupCoordinator is provided the metadata it needs.

zzbennett committed:
KAFKA-19441: encapsulate MetadataImage in GroupCoordinator/ShareCoordinator

zzbennett committed:
minor cleanup

zzbennett committed:
fix linting

zzbennett committed:
Merge branch 'trunk' of github.com:apache/kafka into lbennett-kafka-19441

zzbennett committed:
fix linting

dajac commented on coordinator-common/src/main/java/org/apache/kafka/coordinator/common/runtime/CoordinatorMetadataImage.java:
```diff
@@ -0,0 +1,145 @@
+    default Optional<TopicMetadata> topicMetadata(String topicName) {
+        var topicId = topicId(topicName);
+        return topicId.isEmpty() ? Optional.empty() : topicMetadata(topicId.get());
+    }
```
nit: Should we provide the default implementation for the topic id version in order to be consistent with `partitionCount`?

dajac commented on coordinator-common/src/main/java/org/apache/kafka/coordinator/common/runtime/CoordinatorMetadataImage.java:
```diff
@@ -0,0 +1,145 @@
+
+    CoordinatorMetadataDelta emptyDelta();
+
+    Long version();
```
nit: Could we use `long` here?

dajac commented on coordinator-common/src/main/java/org/apache/kafka/coordinator/common/runtime/CoordinatorMetadataImage.java:
```diff
@@ -0,0 +1,145 @@
+
+        default Set<Integer> partitionSet() {
+            return IntStream.range(0, partitionCount()).boxed().collect(Collectors.toSet());
+        }
```
This is annoying. Let me check where we still use this.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorService.java:
```diff
@@ -1921,14 +1921,14 @@ private CompletableFuture<DeleteShareGroupOffsetsResponseData> handleDeleteShare
                     new DeleteShareGroupOffsetsResponseData.DeleteShareGroupOffsetsResponseTopic()
                         .setTopicId(topicData.topicId())
-                        .setTopicName(metadataImage.topics().topicIdToNameView().get(topicData.topicId()))
+                        .setTopicName(metadataImage.topicName(topicData.topicId()).orElse("<UNKNOWN>"))
```
We cannot return `<UNKNOWN>`. What was returned with the previous implementation? `null`?

zzbennett replied:
Yea before we were returning `null`. I had seen some other places where we were setting topic name to `<UNKNOWN>` e.g. https://github.com/apache/kafka/blob/trunk/group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java#L4924 but to maintain the existing behavior we can just keep it as null

lbradstreet replied:
Hmm, I was going to say that an error code should be set, but this is the error case anyway. I guess if the current behavior is to return null then I guess we should still return null.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorService.java:
```diff
@@ -1921,14 +1921,14 @@ private CompletableFuture<DeleteShareGroupOffsetsResponseData> handleDeleteShare
                 successTopics.put(
                     topicData.topicId(),
-                    metadataImage.topics().topicIdToNameView().get(topicData.topicId())
+                    metadataImage.topicName(topicData.topicId()).orElse("<UNKNOWN>")
```
ditto. In this case, we can return ZERO_UUID.

zzbennett replied:
this is setting the topic name here so I'm not sure if ZERO_UUID makes sense

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorService.java:
```diff
@@ -2139,7 +2139,7 @@ public void onPartitionsDeleted(
         // with the deleted topics.
         Set<Uuid> topicIds = topicPartitions.stream()
-            .map(tp -> metadataImage.topics().getTopic(tp.topic()).id())
+            .map(tp -> metadataImage.topicId(tp.topic()).orElse(null))
```
ZERO_UUID?

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:
```diff
@@ -5725,24 +5717,23 @@ public void onNewMetadataImage(MetadataImage newImage, MetadataDelta delta) {
+                allGroupIds.addAll(groupsSubscribedToTopic(topicName));
+            }));
+        topicsDelta.deletedTopicIds().forEach(topicId ->
+            delta.image().topicName(topicId).ifPresent(topicName -> {
```
Why do you use `delta.image()` here and `metadataImage` above?

zzbennett replied:
Hm not sure. That is how the code was before but it makes more sense to use the `metadataImage`

zzbennett replied:
Just realized why it is using `delta.image()` thanks to the unit test coverage. The `metadataImage` doesn't contain the deleted topics, only the `delta.image` contains the deleted topics

zzbennett replied:
Added a javadoc to clarify that the image returned from `delta.image()` needs be a snapshot of the metadata _before_ the delta occurred

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/Utils.java:
```diff
@@ -398,38 +390,35 @@ public static long computeGroupHash(Map<String, Long> topicHashes) {
+            .putString(topicMetadata.name())
+            .putInt(topicMetadata.partitionCount());
+
+        Map<Integer, List<String>> racks = topicMetadata.partitionRacks();
```
I don't like this because generating `Map<Integer, List<String>>` is not necessary. Could we change the interface to take the partition index as an argument and return the rack for that partition?

zzbennett replied:
Makes sense, that does make it a bit cleaner

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/Utils.java:
```diff
@@ -398,38 +390,35 @@ public static long computeGroupHash(Map<String, Long> topicHashes) {
-                    broker.rack().ifPresent(racks::add);
-                }
-            }
+            List<String> partitionRacks = new ArrayList<>(racks.get(i));
+            // topicMetadata returns an unmodifiable list
+            Collections.copy(partitionRacks, racks.get(i));
```
This is a tad annoying too because we already allocate a new ArrayList when partitionRacks is called. For performance reasons, I wonder if we should not return an unmodifiable list so we can reuse it here.

This part is performance sensitive. We have a jmh benchmark that we may want to run to ensure that we don't regress here.

zzbennett replied:
Sure I can make it a modifiable list. Making it unmodifiable is probably overkill since, like you said, we already make a copy in the `partitionRacks` method. If we avoid the extra copy, then the new implementation should be equivalent to the old but I'll double check with the jmh benchmarks

zzbennett replied:
I ran the group coordinator jmh benchmarks locally a couple times and didn't see any statistically significant variations in performance

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/modern/SubscribedTopicDescriberImpl.java:
```diff
@@ -66,22 +64,18 @@ public int numPartitions(Uuid topicId) {
+        }
+
+        CoordinatorMetadataImage.TopicMetadata topicMetadata = topicMetadataOp.get();
+        List<String> racks = topicMetadata.partitionRacks().get(partition);
```
This call is pretty nasty because it generate the entire map just to return the racks of a single partition and it does so every time racksForPartition is called. We should definitely consider my suggestion for the interface.

zzbennett replied:
yea agreed...

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/modern/SubscribedTopicDescriberImpl.java:
```diff
@@ -66,22 +64,18 @@ public int numPartitions(Uuid topicId) {
+        CoordinatorMetadataImage.TopicMetadata topicMetadata = topicMetadataOp.get();
+        List<String> racks = topicMetadata.partitionRacks().get(partition);
+        if (racks == null) {
+            return Collections.emptySet();
```
nit: Set.of().

dajac commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java:
```diff
@@ -477,7 +482,7 @@ public void testMemberJoinsEmptyConsumerGroup() {
         GroupMetadataManagerTestContext context = new GroupMetadataManagerTestContext.Builder()
             .withConfig(GroupCoordinatorConfig.CONSUMER_GROUP_ASSIGNORS_CONFIG, List.of(assignor))
-            .withMetadataImage(metadataImage)
+            .withMetadataImage(new KRaftCoordinatorMetadataImage(metadataImage))
```
I am debating whether we should update MetadataImageBuilder to return a CoordinatorMetadataImage instead of wrapping it everywhere. Have you considered it?

zzbennett replied:
I actually did at some point add a method to the MetadataImageBuilder to return a CoordinatorMetadataImage but I must have given up updating the test to use the new method in all the places where it could. This test is really massive 😅

dajac commented on coordinator-common/src/main/java/org/apache/kafka/coordinator/common/runtime/CoordinatorMetadataImage.java:
```diff
@@ -0,0 +1,145 @@
+
+    Optional<TopicMetadata> topicMetadata(Uuid topicId);
+
+    boolean shareGroupsEnabled();
```
This feels weird in the interface. I wonder whether we could expose the features instead. It seems useful for other needs too.

zzbennett replied:
Yea it is weird. The share group coordinator relies on it in its `onNewMetadataImage` method https://github.com/apache/kafka/blob/trunk/share-coordinator/src/main/java/org/apache/kafka/coordinator/share/ShareCoordinatorService.java#L1102 

I just realized though that this particular instance of `onNewMetadataImage` is not part of any of the other coordinator related interfaces though, so I can easily just add the FeatureImage as an extra arg to the `ShareCoordinator` interface instead of handling it through the `CoordinatorMetadataImage`.

AndrewJSchofield replied:
The thing is that the decision whether share groups are enabled depends upon a config and a feature. Either can be used.

AndrewJSchofield replied:
@zzbennett I wonder what your thoughts are on this.

dajac reviewed (commented):
@zzbennett Thanks for the patch. I just made a pass and I left some high level comments. I am not entirely happy with the interface yet.

zzbennett committed:
PR feedback

zzbennett commented:
@dajac thanks for your review. I think I've addressed all your comments in https://github.com/apache/kafka/pull/20061/commits/e66548ec5f941f3f27ee7f9becda8f55e0693809

dajac commented on coordinator-common/src/main/java/org/apache/kafka/coordinator/common/runtime/CoordinatorMetadataImage.java:
```diff
@@ -0,0 +1,134 @@
+
+        int partitionCount();
+
+        default Set<Integer> partitionSet() {
```
Could we remove it?

zzbennett replied:
sure, I can move it out of the interface to keep the interface more clean. There are only 4 places where we use it so I can move the logic inline in those areas

zzbennett committed:
remove the partitionSet() function from TopicMetadata interface

zzbennett committed:
fix linting

lbradstreet commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorService.java:
```diff
@@ -1921,14 +1921,14 @@ private CompletableFuture<DeleteShareGroupOffsetsResponseData> handleDeleteShare
                     new DeleteShareGroupOffsetsResponseData.DeleteShareGroupOffsetsResponseTopic()
                         .setTopicId(topicData.topicId())
-                        .setTopicName(metadataImage.topics().topicIdToNameView().get(topicData.topicId()))
+                        .setTopicName(metadataImage.topicName(topicData.topicId()).orElse(null))
```
Should you ever get null here? I do see that it matches the existing behavior.

zzbennett replied:
It probably wouldn't ever happen but returning null at least matches the current behavior and avoids throwing an exception if the Optional is empty

lbradstreet commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:
```diff
@@ -485,12 +482,12 @@ GroupMetadataManager build() {
      * The cache for topic hash value by topic name.
      * A topic hash is calculated when there is a group subscribes to it.
-     * A topic hash is removed when it's updated in MetadataImage or there is no group subscribes to it.
+     * A topic hash is removed when it's updated in GroupCoordinatorMetadataImage or there is no group subscribes to it.
```
Should this be CoordinatorMetadataImage?

zzbennett replied:
A yea good catch

zzbennett replied:
Found one other place where it was GroupCoordinatorMetadataImage. At first that was what the class was called, but I renamed it to CoordinatorMetadataImage when I realized it was being shared by the Share Group Coordinator too

lbradstreet commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:
```diff
@@ -8120,56 +8119,58 @@ public Map.Entry<AlterShareGroupOffsetsResponseData, InitializeShareGroupStatePa
-                        .collect(Collectors.toSet()),
-                    currentTimeMs
-                ));
+            metadataImage.topicMetadata(topic.topicName()).ifPresentOrElse(
```
This isn't really the ifPresentOrElse's fault but I do find these if blocks a little unwieldy. I don't really have a great suggestion right now -- it's probably better that we leave the current structure so that the refactoring is still fairly mechanical.

zzbennett replied:
By current structure, you mean the if/else block instead of the `ifPresenetOrElse`? I'm fine with either one. I agree the `ifPresentOrElse` is kind of hard to read when it's a big block of logic. It's probably better for one or two line pieces of logic

zzbennett replied:
Updated it to be a regular `if/else` block

zzbennett committed:
cleanup from PR review

zzbennett committed:
Merge branch 'trunk' of github.com:apache/kafka into lbennett-kafka-19441

zzbennett committed:
Merge branch 'trunk' of github.com:apache/kafka into lbennett-kafka-19441

zzbennett commented on coordinator-common/src/main/java/org/apache/kafka/coordinator/common/runtime/CoordinatorMetadataDelta.java:
```diff
@@ -0,0 +1,72 @@
+
+    CoordinatorMetadataDelta EMPTY = emptyDelta();
+
+    interface TopicsDelta {
```
now I'm thinking of removing TopicsDelta and moving createdTopicsIds/changedTopicIds/deletedTopicsIds to the top level interface because it's kind of redundant to have this sub-class here.

zzbennett committed:
remove redundant TopicsDelta interface

zzbennett committed:
slightly enhance metadata delta tests

zzbennett committed:
add a javadoc

lbradstreet commented on coordinator-common/src/main/java/org/apache/kafka/coordinator/common/runtime/CoordinatorMetadataImage.java:
```diff
@@ -0,0 +1,128 @@
+
+    default Optional<TopicMetadata> topicMetadata(Uuid topicId) {
+        var topicName = topicName(topicId);
+        return topicName.isEmpty() ? Optional.empty() : topicMetadata(topicName.get());
```
This impl is fine, but you could do `return topicName(topicId).flatMap(this::topicMetadata)`

Is there a reason why we made topicMetadata(Uuid topicId) the default impl rather than `topicMetadata(String topicName)`? I would generally rather impls key on topicid.

zzbennett replied:
No particular reason. This is moot now though because I removed `topicName()` from the top-level interface, so the underlying impls would need to implement both the name and Uuid version

lbradstreet commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:
```diff
@@ -8117,7 +8112,7 @@ public List<DeleteShareGroupStateRequestData.DeleteStateData> sharePartitionsEli
                     // we will include that topic in the delete share group state request.
                     List<DeleteShareGroupStateRequestData.PartitionData> partitions = new ArrayList<>();
-                    topicImage.partitions().keySet().forEach(partition ->
+                    IntStream.range(0, topicMetadata.partitionCount()).boxed().collect(Collectors.toSet()).forEach(partition ->
```
This looks like a mechanical refactoring that went a bit wrong. We can do:

```
                    List<DeleteShareGroupStateRequestData.PartitionData> partitions = IntStream.range(0, topicMetadata.partitionCount())
                            .mapToObj(partition -> new DeleteShareGroupStateRequestData
                                    .PartitionData()
                                    .setPartition(partition))
                            .toList();
```
instead.

zzbennett replied:
ah good catch, yea that was some copy pasta from removing the built-in `partitions()` method from the CoordinatorMetadata interface

lbradstreet commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:
```diff
@@ -8165,10 +8160,11 @@ public Map.Entry<AlterShareGroupOffsetsResponseData, InitializeShareGroupStatePa
+            if (topicMetadataOpt.isPresent()) {
+                var topicMetadata = topicMetadataOpt.get();
+                Uuid topicId = topicMetadata.id();
+                Set<Integer> existingPartitions = IntStream.range(0, topicMetadata.partitionCount()).boxed().collect(Collectors.toSet());
```
Generating this set seems pretty overkill too. I think we can just delete the set creation and do something like this?

```
                    if (partition.partitionIndex() < topicMetadata.partitionCount()) {
                        partitions.add(
```

zzbennett replied:
yea that makes more sense

lbradstreet commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorService.java:
```diff
@@ -2139,7 +2139,7 @@ public void onPartitionsDeleted(
         // with the deleted topics.
         Set<Uuid> topicIds = topicPartitions.stream()
-            .map(tp -> metadataImage.topics().getTopic(tp.topic()).id())
+            .map(tp -> metadataImage.topicId(tp.topic()).orElse(Uuid.ZERO_UUID))
```
One thing that I was thinking about with the ZERO_UUID handling cases is that this group coordinator is only supported by clusters that use KRaft, and with KRaft you're guaranteed to have a topicId if one exists. So I'm not even sure there should be any cases where we want to return ZERO_UUID to clients?

zzbennett replied:
I guess there is maybe the theoretical case where the `topicPartitions` that was passed in contains topics that don't exist in the metadata (the comment seems to indicate this won't happen though). But I can just return `null` since that is the existing behavior

lbradstreet commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:
```diff
@@ -2727,18 +2724,17 @@ Map<Uuid, InitMapValue> subscribedTopicsChangeMap(String groupId, Set<String> su
+            metadataImage.topicMetadata(topicName).ifPresent(topicMetadata -> {
+                Set<Integer> alreadyInitializedPartSet = alreadyInitialized.containsKey(topicMetadata.id()) ? alreadyInitialized.get(topicMetadata.id()).partitions() : Set.of();
+                if (alreadyInitializedPartSet.isEmpty() || alreadyInitializedPartSet.size() < topicMetadata.partitionCount()) {
+                    Set<Integer> partitionSet = IntStream.range(0, topicMetadata.partitionCount()).boxed()
```
this is preexisting, but I think we could move the partitionSet calculation into the computeIfAbsent?

zzbennett replied:
makes sense to me

lbradstreet commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/modern/ModernGroup.java:
```diff
@@ -367,18 +366,15 @@ public void setMetadataHash(long metadataHash) {
-        subscribedTopicNames.keySet().forEach(topicName -> {
-            TopicImage topicImage = metadataImage.topics().getTopic(topicName);
-            if (topicImage != null) {
+        subscribedTopicNames.keySet().forEach(topicName ->
```
Now that we've added the ifPresent in there I'd suggest writing it fully in streams, e.g.

```
        Map<String, Long> topicHash = subscribedTopicNames.keySet().stream()
                .filter(topicName -> metadataImage.topicMetadata(topicName).isPresent())
                .collect(Collectors.toMap(
                        topicName -> topicName,
                        topicName -> topicHashCache.computeIfAbsent(topicName, k -> Utils.computeTopicHash(k, metadataImage))));
```

zzbennett replied:
nice yea that's better

TaiJuWu commented on coordinator-common/src/main/java/org/apache/kafka/coordinator/common/runtime/CoordinatorMetadataDelta.java:
```diff
@@ -0,0 +1,67 @@
+ */
+public interface CoordinatorMetadataDelta {
+
+    CoordinatorMetadataDelta EMPTY = emptyDelta();
```
Could be static variable?

zzbennett replied:
Actually this field is already implicitly static because it is a field defined in an interface https://docs.oracle.com/javase/specs/jls/se7/html/jls-9.html#jls-9.3

lbradstreet commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:
```diff
@@ -4904,24 +4900,20 @@ public CoordinatorResult<Void, CoordinatorRecord> uninitializeShareGroupState(
         for (Uuid topicId : topicIds) {
-            TopicImage topicImage = topicsImage.getTopic(topicId);
-            String topicName = (topicImage != null) ? topicImage.name() : "<UNKNOWN>";
+            String topicName = metadataImage.topicName(topicId).orElse("<UNKNOWN>");
```
We changed away from UNKNOWN elsewhere.

zzbennett replied:
as discussed offline, in order to avoid making functional changes in the PR, I think we decided to keep it as `<UNKNOWN>` for now

zzbennett committed:
PR review feedback

zzbennett committed:
Merge branch 'trunk' of github.com:apache/kafka into lbennett-kafka-19441

zzbennett committed:
fix build

zzbennett commented:
@lbradstreet addressed your feedback in https://github.com/apache/kafka/pull/20061/commits/3383d798a643bba489c71299ca9f2108a8138f4b

zzbennett committed:
Merge branch 'trunk' of github.com:apache/kafka into lbennett-kafka-19441

# Conflicts:
#	group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

lbradstreet reviewed (approved):
LGTM, thanks for the PR!

zzbennett committed:
fix compilation issue in jmh benchmarks

TaiJuWu commented on coordinator-common/src/main/java/org/apache/kafka/coordinator/common/runtime/CoordinatorRuntime.java:
```diff
@@ -2474,37 +2472,37 @@ public void scheduleUnloadOperation(
         // Push an event for each coordinator.
         coordinators.keySet().forEach(tp -> {
-            scheduleInternalOperation("UpdateImage(tp=" + tp + ", offset=" + newImage.offset() + ")", tp, () -> {
+            scheduleInternalOperation("UpdateImage(tp=" + tp + ", offset=" + newImage.version() + ")", tp, () -> {
```
Should we change the `offset=` to `version=`?
There are many similar issues in this PR .

zzbennett replied:
I will update `offset=` to `version=` here, and update the log statements in this method. I also found a few comments in `GroupMetadataManager.onNewMetadataImage` that reference "offset" instead of "version" so I will update those as well

TaiJuWu commented on coordinator-common/src/main/java/org/apache/kafka/coordinator/common/runtime/KRaftCoordinatorMetadataDelta.java:
```diff
@@ -0,0 +1,82 @@
+    @Override
+    public CoordinatorMetadataImage image() {
+        return new KRaftCoordinatorMetadataImage(metadataDelta.image());
+    }
```
This method can be moved to above `toString`.

TaiJuWu commented on coordinator-common/src/main/java/org/apache/kafka/coordinator/common/runtime/KRaftCoordinatorMetadataImage.java:
```diff
@@ -0,0 +1,149 @@
+            }
+        }
+    }
+}
```
Could you add new line for all files?

TaiJuWu reviewed (commented):
Thank for this patch, leave some comments, PTAL.

zzbennett committed:
Merge branch 'trunk' of github.com:apache/kafka into lbennett-kafka-19441

zzbennett committed:
more PR feedback

TaiJuWu commented on coordinator-common/src/main/java/org/apache/kafka/coordinator/common/runtime/CoordinatorMetadataImage.java:
```diff
@@ -0,0 +1,99 @@
+ * Implementations should be thread-safe and immutable.
+ */
+public interface CoordinatorMetadataImage {
+    static CoordinatorMetadataImage EMPTY = emptyImage();
```
static is redunant.

TaiJuWu commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/streams/topics/EndpointToPartitionsManager.java:
```diff
@@ -38,7 +38,7 @@ private EndpointToPartitionsManager() {
                                                                                               final StreamsGroupHeartbeatResponseData.Endpoint responseEndpoint,
                                                                                               final StreamsGroup streamsGroup,
-                                                                                              final MetadataImage metadataImage) {
+                                                                                              final CoordinatorMetadataImage metadataImage) {
```
Have you considered renaming this parameter to coordinatorMetadataImage to avoid confusion with original `metadataImage`?

zzbennett replied:
I was considering that. But since the `CoordinatorMetadataImage` should be used in the same way as the `MetadataImage`, and there should not be a use case where both images are being used for different purposes, I felt that keeping the `metadataImage` name better conveys how the image should be used.

TaiJuWu commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/streams/topics/InternalTopicManager.java:
```diff
@@ -50,16 +49,16 @@ public class InternalTopicManager {
+     * @param logContext    The log context.
+     * @param metadataHash  The metadata hash of the group.
+     * @param topology      The topology.
+     * @param metadataImage The topics image.
```
How about `the coordinator metadata image contain topics`?

zzbennett replied:
I just updated it to The metadata image, since that is what is used in most of the other javadocs

TaiJuWu reviewed (commented):
Thanks for this patch ,leave some comments.
Overall LGTM.

zzbennett committed:
more PR feedback

TaiJuWu reviewed (approved):
LGTM, thanks for this patch.

zzbennett committed:
Merge branch 'trunk' of github.com:apache/kafka into lbennett-kafka-19441

zzbennett committed:
Merge branch 'trunk' of https://github.com/apache/kafka into lbennett-kafka-19441

zzbennett committed:
empty commit to retrigger CI
